operator: Add Keepalived Controller

TeddyAndrieux opened the pull request:
- Update ClusterConfig CRD
- Add some utils to manage Objects
- Add Go variable in operator with MetalK8s version
- Add skeleton for "sub reconciler"
- Handle MetalK8s VIP Namespace creation/update/deletion
- Handle MetalK8s VIP High-Level ConfigMap creation/update/deletion
- Handle MetalK8s VIP DaemonSet creation/update/deletion
- Add (and update) some status on ClusterConfig CR
- Add some events on ClusterConfig CR

NOTE: For reviewers, you should review commits one by one

NOTE: Some features will be handled later (in other PRs)
- Spread constraints on the VIPs
- CNI reconfiguration (likely once we will have migrated to Tigera operator, to manage the CNI)
- Ingress server certificate "reconfiguration"

---

NOTE: This PR is a work in progress, still some stuff todo

- Add Go unit tests => in progress
- Add tests about VIPs management
- Add documentation
- Add changelog entry

TeddyAndrieux committed:
operator: Add VirtualIPPool specification in ClusterConfig CRD

Re-generate go code and manifests using:
```
make generate
make manifests
make metalk8s
```

TeddyAndrieux committed:
operator: Move the reconciler `Client` as a field of the Reconciler

So that it allow to pass this `Client` to some other function, struct,
ect

TeddyAndrieux committed:
operator: Add skeleton for VirtualIP "sub reconciler"

TeddyAndrieux committed:
operator: Allow the operator to own `Namespace`

```
make manifests
make metalk8s
```

TeddyAndrieux committed:
operator: Add Go variable with the MetalK8s Version

TeddyAndrieux committed:
operator: Add ObjectHandler struct to manage Object manipulation

This ObjectHandler allow to use a single function to Create, Update and
Delete several objects, it also set some standard metadata like Labels
and Owner.

It use the same "mutate" function logic as the
`controllerutil.CreateOrUpdate` function

TeddyAndrieux committed:
operator: Move reconcile result output in an utils module

TeddyAndrieux committed:
operator: Handle MetalK8s VIP Namespace creation, update and deletion

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/metalk8s/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @teddyandrieux

* @gdemonet

* @chengyanjin

* @jbwatenbergscality

* @lucieleonard

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @jbwatenbergscality

* @chengyanjin

* @thomasdanan

* @gdemonet

* @teddyandrieux

TeddyAndrieux committed:
operator: Allow the operator to own `ConfigMap`

```
make manifests
make metalk8s
```

TeddyAndrieux committed:
operator: Allow the operator to Get `Node` objects

```
make manifests
make metalk8s
```

TeddyAndrieux committed:
operator: Mark the "SpreadConstraints" as Not supported Yet

TeddyAndrieux committed:
operator: Add Struct to manage the High Level Config content

TeddyAndrieux committed:
operator: Handle MetalK8s VIP ConfigMaps creation, update and deletion

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @chengyanjin

TeddyAndrieux committed:
operator: Allow the operator to own `DaemonSet`

```
make manifests
make metalk8s
```

TeddyAndrieux committed:
operator: Add Standard mutate for DaemonSet in ObjectHandler

TeddyAndrieux committed:
operator: Handle MetalK8s VIP DaemonSets creation, update and deletion

TeddyAndrieux committed:
operator: Add simple function to GetImageName

TeddyAndrieux committed:
salt: Add Virtual IPs from ClusterConfig CR as portmap IPs

Since, today, the CNI config and Ingress server certificate are managed
by Salt and we want to use the ClusterConfig Virtual IPs to access the
Workload Plane Ingress, retrieve those configured Virtual IPs in Salt so
that we use them to setup CNI and Ingress server certificate

TeddyAndrieux committed:
operator: Add checksum annotation for VIP DaemonSet config

TeddyAndrieux committed:
operator: Add ClusterConfig status conditions

Re-generate go code and manifests using:
```
make generate
make manifests
make metalk8s
```

TeddyAndrieux committed:
operator: Add a global "Configured" condition status

TeddyAndrieux committed:
operator: Add "Configured" condition status for VirtualIPs

TeddyAndrieux committed:
operator: Add some events in VirtualIP reconciler

TeddyAndrieux committed:
operator: Allow the operator to manage `Events`

```
make manifests
make metalk8s
```

gdemonet commented on operator/pkg/controller/utils/reconcile.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// Get the Return for the main Reconcile function
+func (r ReconcilerResult) GetReturn() (ctrl.Result, error) {
```
nit: should likely be called `GetResult`, since that's the type name of the method's return value.

TeddyAndrieux replied:
True but this function return the Result + error, but yes maybe GetResult is clearer, I will change it

gdemonet commented on operator/pkg/controller/utils/reconcile.go:
```diff
@@ -0,0 +1,70 @@
+	return newReconcilerResult(delayRequeue, nil)
+}
+
+// End the Reconciliation
```
nit: no need for a capital R :wink:

gdemonet commented on operator/pkg/controller/utils/object.go:
```diff
@@ -0,0 +1,160 @@
+
+	labelComponentName = "app.kubernetes.io/component"
+
+	labelVersionName = "metalk8s.scality.com/version"
```
There is a standard label key for this as well (`app.kubernetes.io/version`), IMO we should use it in addition to our own

TeddyAndrieux replied:
Right, that makes sense, I will add it.

Edit: Hmm I think we want to keep the `app.kubernetes.io/version` for the version of the component itself (so in our case for keepalived since we use the MetalK8s it will be the same but, it's not always the same)
But still I will add it just it will not always match the MetalK8s version

gdemonet commented on operator/pkg/controller/utils/object.go:
```diff
@@ -0,0 +1,160 @@
+	labels map[string]string
+}
+
+func NewObjectHandler(instance *metalk8sv1alpha1.ClusterConfig, client client.Client, scheme *runtime.Scheme, logger logr.Logger, component string) *ObjectHandler {
```
Shouldn't we always document public functions (and types, I guess)?
In this specific case, I'm not sure what `component` is used for, and I only know what an ObjectHandler does because I read the commit message :)

TeddyAndrieux replied:
Ok, make sense

gdemonet commented on operator/pkg/controller/virtualip/hl_config.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func (c *HLConfig) isSupportedApiVersion() bool {
+	for i := range supportedApiVersion {
+		if c.ApiVersion == supportedApiVersion[i] {
+			return true
+		}
+	}
```
Why the index-based loop here, while we usually just loop through items?

TeddyAndrieux replied:
TBH, no good reason, I will change it

gdemonet commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -69,7 +83,213 @@ func (r *VirtualIPReconciler) Reconcile(ctx context.Context) utils.ReconcilerRes
+
+	poolsName := []string{}
+
+	for pName := range pools {
```
Shouldn't this be `for _, pName := range pools {` ?

TeddyAndrieux replied:
No, because `pools` is a map and here I iterate on the map "keys" (and I do not need any information from the "values")

gdemonet commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -69,7 +83,213 @@ func (r *VirtualIPReconciler) Reconcile(ctx context.Context) utils.ReconcilerRes
+	objsToDelete := []client.Object{}
+	objsToUpdate := []client.Object{}
+
+	poolsName := []string{}
```
nit: `poolNames` :wink:

gdemonet commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -0,0 +1,39 @@
+
+type VirtualIPReconciler struct {
+	utils.ObjectHandler
+}
```
This is black magic to me, do you have some reading material I could get my hands on to understand this better? IIUC, by defining this "anonymous" field in a struct, the struct inherits the public methods from its child...? e.g. the `CreateOrUpdateOrDelete` method from the commit before?

TeddyAndrieux replied:
I do not have any specific link for it, but yes exactly, it's like inheritance

gdemonet commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -15,6 +19,9 @@ import (
 type VirtualIPReconciler struct {
 	utils.ObjectHandler
+
+	usedVRID map[int]bool
```
Could be worth adding a comment here, explaining that we need this cache to ensure we don't reuse VRIDs, even for different pools

TeddyAndrieux replied:
Fair

gdemonet commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -293,3 +309,80 @@ func (r *VirtualIPReconciler) mutateConfigMap(obj *corev1.ConfigMap) error {
+		},
+	}
+
+	// If we have more than 1 container we clean up the containers list
```
Guess that's fine for now, but why do we need to do this? Can't we just select by name? I'm thinking we might want to let admins add some sidecar, but not sure... Anyway, that's nitpicking, no need to change anything until the need comes up.

TeddyAndrieux replied:
We could use the name to retrieve the right container, but, to me, we manage this DaemonSet so the user is not supposed to change it, if we want to support sideCars we will want to expose it in our ClusterConfig

gdemonet commented on operator/api/v1alpha1/clusterconfig_types.go:
```diff
@@ -84,6 +87,57 @@ type ClusterConfig struct {
+	condition := metav1.Condition{
+		Type:               kind,
+		Status:             status,
+		ObservedGeneration: c.Generation,
```
Oh! Is this "standard", i.e. always part of `metav1.Condition`? I'm thinking we may have "over-engineered" https://github.com/scality/zenko-operator/pull/328 (cc @williamlardier) while having this available in each condition allows us to know which one was updated after seeing the CR update... :thinking: Anyway, the current approach should work as well :innocent: but it's good to know!

williamlardier replied:
Hey, yes we thought about that for Zenko as well, but we wanted to have it as a standalone value, similar to the deployments, so we can easily check its value regardless of the conditions. But maybe this can be improved, depending on what we want to achieve at the product level.

TeddyAndrieux replied:
> Oh! Is this "standard", i.e. always part of `metav1.Condition`?

Indeed, and from what I see (but I did not tested myself) the `kubectl wait --for=condition` honor it by default which means that it will wait for the `ObservedGeneration` to match the object `Generation`

williamlardier replied:
cc @francoisferrand do you think this could be helpful in a Zenko context?

gdemonet commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -157,17 +202,32 @@ func (r *VirtualIPReconciler) Reconcile(ctx context.Context) utils.ReconcilerRes
 
 	// TODO: Reconfigure CNI + re-generate the Ingress server certificate
 
+	r.updateCondition(metav1.ConditionTrue, "Configured", "All objects properly configured")
```
We stop here? What about checking that the keepalived DaemonSets are actually ready? I don't think it's actually needed atm, and it's even probably not as useful as one might think (we are more interested in knowing the status of the VIPs themselves than that of the Pods), but just wanted to make that clear... It's fine if we say that for now, our status only cares about "K8s objects being properly configured", and once we add monitoring (with our sidecar "API"), we'll add more info about the availability of each VIP.

TeddyAndrieux replied:
I was wondering about it as well, it was simple to only handle "configuration" but I think I will wait for the DaemonSet pods to be ready as well

gdemonet reviewed (commented):
As usual, really good stuff :100: 

I was a bit confused at first by the "magic update/create/delete with a generic mutate handler", but it does make the code less redundant and keeps things simpler when considering a single object type to mutate.

Otherwise, just had a few questions and requests here and there, but nothing blocking or of significant impact :wink:

JBWatenbergScality commented on operator/pkg/controller/virtualip/controller.go:
```diff
@@ -0,0 +1,474 @@
+	}
+	container := &obj.Spec.Template.Spec.Containers[0]
+
+	container.Name = "keepalived"
```
I'm wondering wether we should abstract away this controller from keepalived and not take assumption that we are relying on it for the implementation. We could introduce an intermediate layer "VirtualIpProvider" or something similar which could then be implemented using keepalived, metallb or whatever else if a new virtualip implementation suits better our needs later on, this would ease keepalived replacement

TeddyAndrieux replied:
Hmm yes, we can do this, not sure it's really worth but ... may make sense.

Edit: The fact is that even the "HLConfig" is quite linked to keepalived, we also need a "global" knowledge of all the Virtual Router IDs used.
If one day we change the "Virtual IP Provider" to me the logic will be quite different, I'm not sure it's really worth building an intermediate layer today, that will be over-engineer to be able to support potential other providers (or that will need some tuning in the future to support "this" other provider).
If we change the "Virtual IP Provider", to me, the only logic we can keep from this "controller" is the namespace "management" and .... that's pretty much it, all the re-usable logic is part of the "ObjectHandler" object to me

TeddyAndrieux commented:
Close this PR as it get reworked and merged by #3864
